wp/lib/bap_wp/src/precondition.ml

```diff
@@ -398,12 +399,21 @@ let subst_fun_outputs ?tid_name:(tid_name = "") ~inputs:(inputs : Var.t list)
       Env.add_call_pred env sub_to) in
   Constr.substitute post subs_from subs_to, env
 
+(* Creates fresh names for [vars] found in the constraint. Optionally takes in
+   a function that maps the original variable name to the fresh name. Defaults
+   to [fresh_var]. *)
+let freshen ?(name = Format.sprintf "fresh_%s") (constr : Constr.t)
+    (env : Env.t) (vars : Var.Set.t) : Constr.t * Env.t =
+  Var.Set.fold vars ~init:(constr, env) ~f:(fun (constr, env) v ->
+      let z3_v, env = Env.get_var env v in
+      let name = name (Expr.to_string z3_v) in
+      let fresh = Env.new_z3_expr ~name env (Var.typ v) in
+      Constr.substitute_one constr z3_v fresh, Env.add_call_pred env fresh)
```

Comment:
This gonna sound silly, but any reason we don't do them all at once?

Reply:
`Env.add_call_pred` was added a long time after `Constr.substitute_one`. It was added so that variables that are introduced during the WP would also be printed out in the model (i.e. in the function specs and here in the loop invariant checker). I'm not sure we want to combine the two because there are many occasions where we substitute an expression without introducing a new variable.

Reply:
I just mean use `Constr.substitute` on all the vars at once, instead of folding and doing a bunch of `Constr.substitute_one`.

Maybe we optimize this at the end when building the Z3 constraint?

Reply:
Ooh! I have definitely use `Constr.substitute` instead.